src/magnifier.c

```diff
@@ -0,0 +1,300 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <assert.h>
+#include <wlr/types/wlr_output.h>
+#include "magnifier.h"
+#include "labwc.h"
+#include "theme.h"
+#include "common/macros.h"
+
+bool magnify_on;
+double mag_scale = 0.0, mag_increment = 0.0;
+
+#define CLAMP(in, lower, upper) MAX(MIN(in, upper), lower)
+
+void
+magnify(struct output *output, struct wlr_buffer *output_buffer, struct wlr_box *damage)
+{
+	int width, height;
+	double x, y;
+	struct wlr_box border_box, dst_box;
+	struct wlr_fbox src_box;
+	bool fullscreen = false;
+
+	/* Reuse a single scratch buffer */
+	static struct wlr_buffer *tmp_buffer = NULL;
+	static struct wlr_texture *tmp_texture = NULL;
+
+	/* TODO: This looks way too complicated to just get the used format */
+	struct wlr_drm_format wlr_drm_format = {0};
+	struct wlr_shm_attributes shm_attribs = {0};
+	struct wlr_dmabuf_attributes dma_attribs = {0};
+	if (wlr_buffer_get_dmabuf(output_buffer, &dma_attribs)) {
+		wlr_drm_format.format = dma_attribs.format;
+		wlr_drm_format.len = 1;
+		wlr_drm_format.modifiers = &dma_attribs.modifier;
+	} else if (wlr_buffer_get_shm(output_buffer, &shm_attribs)) {
+		wlr_drm_format.format = shm_attribs.format;
+	} else {
+		wlr_log(WLR_ERROR, "Failed to read buffer format");
+		return;
+	}
+
+	/* Fetch scale-adjusted cursor coordinates */
+	struct server *server = output->server;
+	struct theme *theme = server->theme;
+	struct wlr_cursor *cursor = server->seat.cursor;
+	double ox = cursor->x;
+	double oy = cursor->y;
+	wlr_output_layout_output_coords(server->output_layout, output->wlr_output, &ox, &oy);
+	ox *= output->wlr_output->scale;
+	oy *= output->wlr_output->scale;
+	if (theme->mag_width == -1 || theme->mag_height == -1) {
+		fullscreen = true;
+	}
+	if ((ox < 0 || oy < 0 || ox >= output_buffer->width || oy >= output_buffer->height)
+		&& fullscreen) {
+		return;
+	}
+
+	if (mag_scale == 0.0) {
+		mag_scale = theme->mag_scale;
+	}
+	if (mag_increment == 0.0) {
+		mag_increment = theme->mag_increment;
+	}
```

Comment:
This causes an issue on the first `ZoomIn` action, it will set the mag_scale to `1.0` (`1.0 + 0.0`).
Can we use `theme->mag_increment` everywhere and remove the local var for `mag_increment`?

Reply:
Ah yes - we can do now, because I am passing the server to the action functions, so they have access to the theme directly. Hang on...

Reply:
Done